openfisca_core/taxbenefitsystems.py

```diff
@@ -401,3 +402,23 @@ def get_variables(self, entity = None):
                 for variable_name, variable in self.variables.items()
                 if variable.entity == entity
                 }
+
+    def clone(self):
+        new = empty_clone(self)
+        new_dict = new.__dict__
+
+        for key, value in self.__dict__.items():
+            if key not in ('parameters', '_parameters_at_instant_cache', 'variables', 'open_api_config'):
+                new_dict[key] = value
+
+        new_dict['parameters'] = self.parameters.clone()
+        new_dict['_parameters_at_instant_cache'] = {}
+        new_dict['variables'] = self.variables.copy()
+        new_dict['open_api_config'] = self.open_api_config.copy()
+        return new
+
+    def entities_plural(self):
```

Comment:
> I don't really know what to think of the multiplications of tiny methods like this. 
> If you have some references, I'd like to read more about the benefits of this practice

Reply:
@fpagnoux : It's a consequence of the application of the Demeter's law, to access only the objects that are directly accessible from each class.

Reply:
Yes.

There's a further benefit too: as we prevent SimulationBuilder from directly accessing the internal implementation of TaxBenefitSystem, we make clear what the contract is for the dependency between SimulationBuilder and TaxBenefitSystem.

Eventually we'll notice that whenever we call `tax_benefit_system.something()` the something() has to do with the singular or plural of entities. The sum of these `something()`s defines an interface. This interface is all that we have to pass to SimulationBuilder rather than the entire TaxBenefitSystem. (The TaxBenefitSystem is needed by Simulation, but it could be passed to it by a different route.)

Suppose we call this interface EntityDictionary or something like that. Then to test SimulationBuilder we don't have to instantiate a full TaxBenefitSystem. We can instantiate the much smaller EntityDictionary. And we can instantiate it as a small fixture internal to the test, rather than having to take the entire country template. We get smaller, faster, more decoupled tests.

But the benefit also extends to the objects - we now have a way of assembling a TaxBenefitSystem in smaller steps: first we build an EntityDictionary, then we do the rest. Of course "the rest" can probably be split up as well. TaxBenefitSystem could be thought of as the combination of a VariableRepository, a ParameterRepository and possibly other things. 

All of these smaller objects are easier to unit test. We don't need to load the variables when we want to test parameters, and vice versa.

References: Martin Fowler's _Refactoring_ (2nd edition just came out), Bob Martin's _Clean Code_, Michael Feather's _Working Effectively With Legacy Code_, Sandi Metz's _Practical Object Oriented Design_.

Reply:
Ok, thanks a lot for taking the time to explain and for the references!